**By Using Single Value Decomposition**

In [3]:
import pandas as pd

In [8]:
data = pd.read_csv('ratings.csv')
data.head()

book_id  user_id  rating
0        1      314       5
1        1      439       3
2        1      588       5
3        1     1169       4
4        1     1185       4

In [4]:
pip install scikit-surprise

     |████████████████████████████████| 11.8MB 257kB/s 
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.1-cp36-cp36m-linux_x86_64.whl size=1670922 sha256=f560f01fa256a418a09ae09cd4e56ada84595c1c01e407155acb20f6fc0f2842
  Stored in directory: /root/.cache/pip/wheels/78/9c/3d/41b419c9d2aff5b6e2b4c0fc8d25c538202834058f9ed110d0
Successfully built scikit-surprise


In [9]:
from surprise import Reader, Dataset, SVD
from surprise.model_selection import cross_validate
reader = Reader()

In [10]:
df = Dataset.load_from_df(data[['book_id', 'user_id', 'rating']], reader)

In [11]:
svd = SVD()
cross_validate(svd, df, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8422  0.8437  0.8439  0.8435  0.8447  0.8436  0.0008  
MAE (testset)     0.6593  0.6602  0.6596  0.6593  0.6595  0.6596  0.0004  
Fit time          46.50   47.45   47.04   47.33   46.80   47.02   0.35    
Test time         2.18    2.19    2.17    2.22    2.22    2.20    0.02    


{'fit_time': (46.502410650253296,
  47.44561195373535,
  47.03673219680786,
  47.332595348358154,
  46.79761791229248),
 'test_mae': array([0.65925383, 0.66021552, 0.65962798, 0.65925661, 0.65952612]),
 'test_rmse': array([0.84219014, 0.84366262, 0.84392678, 0.84353641, 0.84466355]),
 'test_time': (2.1839609146118164,
  2.190577268600464,
  2.1721339225769043,
  2.2228810787200928,
  2.2152328491210938)}

In [12]:
trainset = df.build_full_trainset()
svd.fit(trainset)

In [14]:
data[data['user_id'] == 1]

book_id  user_id  rating
117889     1180        1       4
488112     4893        1       3
625717     6285        1       4

In [21]:
svd.predict(1, 182, 3)

Prediction(uid=1, iid=182, r_ui=3, est=4.676406189811783, details={'was_impossible': False})

For book with ID 182 , we get an estimated prediction of 4.6764, According to the user 1. 